onnxruntime/core/providers/cpu/math/einsum_auxiliary_ops.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// This module hosts implementations and thin wrappers over other onnx operator implementations
+// that will be called from within the Einsum operator implementation
+
+#pragma once
+
+#include "core/common/common.h"
+#include "core/framework/allocator.h"
+#include "core/util/math.h"
+#include "core/providers/cpu/tensor/transpose.h"
+#include "core/providers/cpu/reduction/reduction_ops.h"
+
+#include <vector>
+
+namespace onnxruntime {
+
+namespace EinsumOp {
+
+// Thin wrapper over the Transpose op
+Tensor Transpose(const Tensor& input, const std::vector<size_t>& permutation, AllocatorPtr allocator);
+
+// Creates a "reshaped view" for the same tensor (i.e.) mutates the shape for the same tensor
+// We will use it to introduce some "unsqueezed" dims (i.e.) extra dims with dim value as 1
+inline void CreateReshapedView(Tensor& input, const std::vector<int64_t>& new_dims) {
+  input.Reshape(new_dims);
+}
+
```

Comment:
Would it be clearer just to directly call Reshape?

Reply:
It could. But I would prefer to keep this an inlined method (can force the compiler to inline maybe rather than just the hint), documenting why we do it in the header near the implementation might seem easier for others to grasp rather than generic arbitrary calls to .Reshape() spread across the implementation. What do you think ?

Reply:
You can't force a compiler to do anything. In C++ 'inline' is only ever a hint. 

---
In reply to: [414340800](https://github.com/microsoft/onnxruntime/pull/3575#discussion_r414340800) [](ancestors = 414340800)

Reply:
Yes, I know. “inline” is just a hint but there are things like _forceinline (or something equivalent for another compiler) that forces the compiler to inline, doesn’t it ?(This method is gone in the refactoring, so no action needed anyway)